Added tracing options

coot opened the pull request:
This is useful for debugging.

mrBliss commented:
Nice! Could you also add one for the ChainDB? It would be here: https://github.com/input-output-hk/cardano-node/blob/b88c29f2ea945fbf6c42d1c232817383c4439899/app/Run.hs#L490 It should use `readableChainDBTracer` by default, but swap to a regular tracer when enabled. This will be very handy when debugging.

coot commented:
Sure, but that one ought to be on by default, isn't it?

mrBliss commented:
> Sure, but that one ought to be on by default, isn't it?

I mean that by default the human-readable tracer should be used, but a full tracer (`showTracing tracer`?) when enabled.

coot commented:
@mrBliss done, I also added logger names, this makes using grep more effective.
Note that I removed the `ChainDB |` prefix, since it is now present in the logger name.

mrBliss commented:
@coot I have been thinking and had another idea:

Instead of sprinkling things like the following all over the code:
```haskell
enableTracer traceFetchDecisions $ withName "FetchDecision" trace
```
We should just make one big record with all the tracers and pass that to the node and node-network setup code:

```haskell
data Tracers m blk = Tracers
  { tracerChainDB        :: !Tracer m (ChainDB.TraceEvent blk)
  , tracerChainSync      :: !Tracer m ..
  , ..
  }
```
In the setup code (i.e. `cardano-node`), we can just forward the right tracers.

Then we can have some predefined values of `Tracers`, one with all `nullTracer`s, one with human-readable output, one with verbose tracers, and a function that uses `TraceOptions` to selectively enable/disable tracers. If a `--(trace-)verbose` is passed, the boolean toggles will apply to the verbose tracer record instead of the (default) human-readable output tracer.

What do you think?

Plan:
* I'll get started and make the necessary changes to `Ouroboros.Consensus.Node` (and related modules):
  - Replace the `String`-based tracer with proper event-type tracers (this has been on my list for a while and will make other things I'm working on easier)
  - Introduce the `Tracers` record and use it
  - Provide a `nullTracers` value
  - Provide a `verboseTracers` value
  - Provide a `humanReadableTracers` value
* Update this PR when merged:
  - Provide a `applyTraceOptions` function
  - ...

mrBliss commented:
See  https://github.com/input-output-hk/ouroboros-network/pull/839

coot commented:
@mrBliss I pushed the change which adds all the traces in one place in `cardano-node`.  I left a few `TODO`s which require modifications in `ouroboros-network`.  This can be merged before changes in `ouroboros-consensus`.

coot commented:
I just saw you're done with your PR, I'll review it after I come back from my lunch.

dcoutts commented on app/Run.hs:
```diff
@@ -545,53 +504,240 @@ mkChainDbArgs TraceOptions {traceChainDB} cfg initLedger registry (CoreNodeId ni
+    -- verbose tracer
+    --
+      tracerChainDB
+        :: !(Tracer IO(WithTip blk (ChainDB.TraceEvent blk)))
```
There's no need to make these strict. It doesn't buy anything I think.

coot replied:
Yes, these are just lambdas, and thus not redexes.

dcoutts commented on app/Run.hs:
```diff
@@ -545,53 +504,240 @@ mkChainDbArgs TraceOptions {traceChainDB} cfg initLedger registry (CoreNodeId ni
+    -- | consensus tracer
+    --
+    , tracerConsensus
+      :: !(Tracer IO String)
```
"String" here is a red flag. Add a TODO.

coot replied:
@mrBliss has just fixed it in `ouroboros-consensus`. It will be fixed today.

coot replied:
I'll do that in a separate PR.

dcoutts commented on app/Run.hs:
```diff
@@ -545,53 +504,240 @@ mkChainDbArgs TraceOptions {traceChainDB} cfg initLedger registry (CoreNodeId ni
+        -> Tracer IO a
+withTip varTip tr = Tracer $ \msg -> do
+    tip <- atomically $ readTVar varTip
+    traceWith (contramap (WithTip tip) tr) msg
```
I think we need a TODO to do this differently. The consensus tracing can return this directly, and importantly can grab it in a way that's consistent with the event. Grabbing it in a separate STM transaction risks reporting different info.

coot replied:
Making this work here was a bit tricky, and moving this to `ouroboros-consensus` would simplify things.  I'll add a TODO.

^ @mrBliss

coot commented:
bors r+

coot commented:
@mrBliss can I merge it? Then I'll be able to finish the `cardano-node` side.

coot commented:
ups, I posted this in the wrong PR, I mean to merge your changes in `ouroboros-consensus`: https://github.com/input-output-hk/ouroboros-network/pull/839

mrBliss commented:
@coot I would have liked to get some feedback on an alternative approach: https://github.com/input-output-hk/cardano-node/pull/83 before merging this one, but I see you summoned bors already.

dcoutts commented:
No problem, we can still consider alternatives from where we are now.

coot commented:
Indeed, and we can split the rather big `Traces` record in two parts, network / consensus.

coot commented:
@mrBliss this PR timed out, anyway it'd prefer to merge it and improve on top of this.  At the moment we cannot advance to the tip of `ouroboros-network` as some other changes needs to be done first.

dcoutts commented:
bors r+
